c/prima.c

```diff
@@ -21,50 +56,127 @@ int lincoa_c(prima_obj calfun, const void *data, const int n, double x[], double
+  }
+  if (opt->npt < 0)
+    opt->npt = 2*n+1;
+  return 0;
+}
+
+void prima_free_options(prima_options * opt)
+{
+  if (opt)
+  {
+    if (opt->_allocated_xl)
+    {
+      free(opt->xl);
+      opt->xl = NULL;
+      opt->_allocated_xl = 0;
+    }
+    if (opt->_allocated_xu)
+    {
+      free(opt->xu);
+      opt->xu = NULL;
+      opt->_allocated_xu = 0;
+    }
+    if (opt->_allocated_nlconstr0)
+    {
+      free(opt->nlconstr0);
+      opt->nlconstr0 = NULL;
+      opt->_allocated_nlconstr0 = 0;
+    }
+  }
+}
+
 /* these functions just call the fortran compatibility layer and return the status code */
-int prima_cobyla(const int m_nlcon, const prima_objcon calcfc, const void *data, const int n, double x[], double *f, double *cstrv, double nlconstr[],
-                 const int m_ineq, const double Aineq[], const double bineq[],
-                 const int m_eq, const double Aeq[], const double beq[],
-                 const double xl[], const double xu[],
-                 int *nf, const double rhobeg, const double rhoend, const double ftarget, const int maxfun, const int iprint)
+int prima_cobyla(const prima_objcon calcfc, const int n, double x[], double *f,
+                 double *cstrv, double nlconstr[],
+                 int *nf, prima_options * opt)
 {
-  int info = 0;
-  cobyla_c(m_nlcon, calcfc, data, n, x, f, cstrv, nlconstr, m_ineq, Aineq, bineq, m_eq, Aeq, beq, xl, xu, nf, rhobeg, rhoend, ftarget, maxfun, iprint, &info);
+  int info = prima_check_options(opt, n);
+  if (info == 0)
+  {
+    // evaluate f0, nlconstr0 if either one is not provided
+    if (opt->f0 == NAN || !opt->nlconstr0)
+    {
+      if (!opt->nlconstr0)
+      {
+        opt->nlconstr0 = (double*)calloc(n, sizeof(double));
+        opt->_allocated_nlconstr0 = 1;
+      }
+      calcfc(x, &(opt->f0), opt->nlconstr0, opt->data);
+    }
+    cobyla_c(opt->m_nlcon, calcfc, opt->data, n, x, f, cstrv, nlconstr,
+            opt->m_ineq, opt->Aineq, opt->bineq, opt->m_eq, opt->Aeq, opt->beq,
+            opt->xl, opt->xu, opt->f0, opt->nlconstr0, nf, opt->rhobeg, opt->rhoend, opt->ftarget, opt->maxfun, opt->iprint, &info);
+    prima_free_options(opt);
+  }
   return info;
 }
 
-int prima_bobyqa(const prima_obj calfun, const void *data, const int n, double x[], double *f, const double xl[], const double xu[],
-                 int *nf, const double rhobeg, const double rhoend, const double ftarget, const int maxfun, const int npt, const int iprint)
+int prima_bobyqa(const prima_obj calfun, const int n, double x[], double *f,
+                 int *nf, prima_options * opt)
 {
-  int info = 0;
-  bobyqa_c(calfun, data, n, x, f, xl, xu, nf, rhobeg, rhoend, ftarget, maxfun, npt, iprint, &info);
+  int info = prima_check_options(opt, n);
+  if (info == 0)
+  {
+    bobyqa_c(calfun, opt->data, n, x, f, opt->xl, opt->xu, nf, opt->rhobeg, opt->rhoend, opt->ftarget, opt->maxfun, opt->npt < 0 ? 2*n+1 : opt->npt, opt->iprint, &info);
+    prima_free_options(opt);
```

Comment:
Calling `prima_free_options` should always be called by the caller of `prima_bobyqa` who has the respomsibilty to create it not by `prima_bobyqa` itself. This also give the opportunity to re-use this structure (for instance to refine the optimization).

The same remark applies for the other algorithms.

The examples have to be modified if you apply the suggested change.

Reply:
I dont think it hurts to reallocate it for each optimization, and that makes the api simpler

Reply:
I changed my mind: lets do what you propose